CodeGeneration/Sources/generate-swiftsyntax/templates/swiftsyntax/SyntaxNodeFile.swift

```diff
@@ -0,0 +1,359 @@
+                        """)
+
+                  }
+                }
+
+                try! InitializerDeclSyntax("public init?<S: SyntaxProtocol>(_ node: S)") {
+                  for choice in child.nodeChoices {
+                    StmtSyntax("""
+                    if let node = node.as(\(raw: choice.typeName).self) {
+                      self = .\(raw: choice.swiftName)(node)
+                      return
+                    }
+                    """)
+                  }
+                  StmtSyntax("return nil")
+                }
+
+                try! VariableDeclSyntax("public static var structure: SyntaxNodeStructure") {
+                  ReturnStmtSyntax(
+                    expression: FunctionCallExprSyntax(callee: ExprSyntax(".choices")) {
+                      TupleExprElementSyntax(
+                        expression: ArrayExprSyntax {
+                          for choice in child.nodeChoices {
+                            ArrayElementSyntax(
+                              expression: ExprSyntax(".node(\(raw: choice.typeName).self)"))
+                          }
+                        }
+                      )
+                    }
+                  )
+                }
+              }
+            }
+          }
+
+          // ==============
+          // Initialization
+          // ==============
+
+          DeclSyntax("public let _syntaxNode: Syntax")
+
+          DeclSyntax("""
+            public init?<S: SyntaxProtocol>(_ node: S) {
+              guard node.raw.kind == .\(raw: node.swiftSyntaxKind) else { return nil }
+              self._syntaxNode = node._syntaxNode
+            }
+            """)
+
+          DeclSyntax("""
+            /// Creates a `\(raw: node.name)` node from the given `SyntaxData`. This assumes
+            /// that the `SyntaxData` is of the correct kind. If it is not, the behaviour
+            /// is undefined.
+            internal init(_ data: SyntaxData) {
+              assert(data.raw.kind == .\(raw: node.swiftSyntaxKind))
+              self._syntaxNode = Syntax(data)
+            }
+            """)
+
+          try! InitializerDeclSyntax("\(raw: node.generateInitializerDecl(optionalBaseAsMissing: false, currentIndentation: "  "))") {
+            let parameters = node.children.map { $0.swiftName }.joined(separator: ", ")
+
+            VariableDeclSyntax(
+              leadingTrivia: [
+                .docLineComment("// Extend the lifetime of all parameters so their arenas don't get destroyed"),
+                .newlines(1),
+                .docLineComment("// before they can be added as children of the new arena."),
+                .newlines(1)
+              ],
+              letOrVarKeyword: .keyword(.let)) {
+                PatternBindingSyntax(
+                  pattern: PatternSyntax("data"),
+                  typeAnnotation: TypeAnnotationSyntax(type: TypeSyntax("SyntaxData")),
+                  initializer: InitializerClauseSyntax(
```

Comment:
Wow, this variable declaration is really deeply nested and long. What do you think about a pattern like the following? I wrote this in GitHub and it might be buggy but I think it gets the idea across.

```swift
let layoutVariable = <lines 164 - 180>
let rawVariable = <lines 183 - 191>
let withExtendedLifetimeParameters = <lines 149 - 151>
let closureSignature = <lines 154 - 161>
let initializer = ExprSyntax("""
  withExtendedLifetime(\(withExtendedLifetimeParameters)) { \(closureSignature) in
    \(layoutVariable)
    \(rawVariable)
    return SyntaxData.forRoot(raw)
  """)
```

---

Also, I just noticed we have a convenience initializer on `VariableDecl` that takes an `initializer` so I t think you don’t need to mess around with the `PatternBindingSyntax`.

Reply:
With the suggested change above we are always initialising and empty array lige `let layout: [RawSyntax?] = []`

Not sure what performance impact it will have.

Reply:
What do mean? AFAICT this should just be moving code around in this file but still  generate the exact same code.

Reply:
Before `<lines 164 - 180>` was inside an if. So when `node.children` was empty is was **not** created. 
Now we get a bit of difference. See `MissingExprSyntax` etc. 

```swift
  public init() {
    // Extend the lifetime of all parameters so their arenas don't get destroyed
    // before they can be added as children of the new arena.
    let data: SyntaxData = withExtendedLifetime((SyntaxArena(), ())) { (arena, _) in 
      let layout: [RawSyntax?] = []
      let raw = RawSyntax.makeEmptyLayout(kind: SyntaxKind.missingExpr, arena: arena)
      return SyntaxData.forRoot(raw)
    }
    self.init(data)
  }
```

Layout is not used